plane/src/typed_unix_socket/client.rs

```diff
@@ -0,0 +1,219 @@
+                                tx.subscribe(),
+                                Arc::clone(&response_map),
+                                event_tx.clone(),
+                            )
+                            .await
+                            .unwrap_or_else(|e| {
+                                tracing::error!("Error handling connection: {}", e);
+                            });
+                        }
+                        Err(e) => {
+                            tracing::error!("Error connecting to server: {}", e);
+                            backoff.wait().await;
+                        }
+                    }
+                }
+            }
+        });
+        Ok(Self {
+            tx,
+            response_map,
+            event_tx,
+        })
+    }
+
+    pub async fn send_request(&self, request: RequestType) -> Result<ResponseType, Error> {
+        let (response_tx, response_rx) = oneshot::channel();
+        let id = Uuid::new_v4();
+        let wrapper = WrappedClientMessageType::Request(IDedMessage {
+            id,
+            message: request,
+        });
+
+        self.response_map.insert(id, response_tx);
+
+        // Wait until there is at least one subscriber
+        let mut backoff = get_quick_backoff();
+        while self.tx.receiver_count() == 0 {
+            tracing::info!("Waiting for a subscriber...");
+            backoff.wait().await;
+        }
+
+        self.tx.send(wrapper)?;
+        let response = response_rx.await?;
+        Ok(response)
+    }
+
+    pub async fn send_message(&self, message: ClientMessageType) -> Result<(), Error> {
+        let wrapper = WrappedClientMessageType::ClientMessage(message);
+
+        // Wait until there is at least one subscriber
+        let mut backoff = get_quick_backoff();
+        while self.tx.receiver_count() == 0 {
+            tracing::info!("Waiting for a subscriber...");
+            backoff.wait().await;
+        }
+
+        self.tx.send(wrapper)?;
+        Ok(())
+    }
+
+    pub fn subscribe_events(&self) -> broadcast::Receiver<ServerMessageType> {
+        self.event_tx.subscribe()
+    }
+}
+
+async fn handle_connection<ClientMessageType, ServerMessageType, RequestType, ResponseType>(
+    stream: UnixStream,
+    mut rx: broadcast::Receiver<WrappedClientMessageType<RequestType, ClientMessageType>>,
+    response_map: Arc<DashMap<Uuid, oneshot::Sender<ResponseType>>>,
+    event_tx: broadcast::Sender<ServerMessageType>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+where
+    ClientMessageType:
+        Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    ServerMessageType:
+        Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    RequestType: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    ResponseType: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
```

Comment:
I gather that `ClientMessageType` and `ServerMessageType` are for "event" messages and `RequestType` / `ResponseType` are for request/response messages. I am inclined to collapse them all into `ClientMessageType` and `ServerMessageType`, and wrap them all in a struct like `IDedMessage` except that makes the ID optional instead of required.

You would still have `send_message` and `send_request`, where `send_message` would give an ID of `None` and `send_request` would set an ID.

My thinking here is that presumably there will be different types of requests/responses, and there's no way at the rust type / serde level to ensure that the response we get back is actually the type we expect given the request we sent, so the code that handles a `ResponseType` response on the client side still needs to extract the correct type of response.

Given that we need to have that complexity live outside of the `TypedUnixSocket` code anyway, I think this would simplify the code here quite a bit.